RecordComponent: Properly handle uninitialized datasets

franzpoeschel opened the pull request:
We currently have a non-functional check during `flush()` that a dataset has been initialized.
The check tests that the `Datatype` is not `UNDEFINED`, but it is never undefined since the constructor initializes it as `CHAR`.

We can use a `std::optional` to have a trustworthy check. This gives users a better feedback if they forget this call.

TODO

- [x] Merge #1278 first
- [x] The newly-added check uncovers that the CoreTests are mostly wrong, since they don't initialize the dataset. Since the CoreTests usually don't need to actually interact with the resulting dataset, they still run fine, but show a warning in the destructor. Maybe fix them.
    ```
    [~Series] An error occurred: Wrong API usage: [RecordComponent] Must specify dataset type and extent before flushing (see RecordComponent::resetDataset()).
    [~Series] An error occurred: Wrong API usage: [RecordComponent] Must specify dataset type and extent before flushing (see RecordComponent::resetDataset()).
    [~Series] An error occurred: Wrong API usage: [RecordComponent] Must specify dataset type and extent before flushing (see RecordComponent::resetDataset()).
    [~Series] An error occurred: Wrong API usage: [RecordComponent] Must specify dataset type and extent before flushing (see RecordComponent::resetDataset()).
    [~Series] An error occurred: Wrong API usage: [RecordComponent] Must specify dataset type and extent before flushing (see RecordComponent::resetDataset()).
    [~Series] An error occurred: Wrong API usage: [RecordComponent] Must specify dataset type and extent before flushing (see RecordComponent::resetDataset()).
    ```
- [ ] Maybe turn this into a warning, and default to `{CHAR, {1}}` instead
- [x] Wait for release 0.15 before merging this. This might uncover bugs in downstream code, and they should have time to fix it.
- [x] Still problems at `myPath` and `structure_test` of CoreTests
- [x] Maybe check if the JSONIOHandlerImpl fix should be applied to other backends, too: The other backends only free their resources in the destructor, but don't flush

Notes:

* This is also useful for things like #1277, since that PR actually gives a dataset with UNDEFINED datatype a meaning (inside TOML templates). Distinguishing between "forgotten to specify" and "explicitly specified as UNDEFINED" is useful here.
  This is an argument in favor of **not** defaulting to `{CHAR, {1}}`
* This seems to uncover some such bugs in our own test suite.

ax3l commented:
This looks good! Anything on the open checks in the PR description we shall discuss tomorrow? :)

franzpoeschel committed:
Use a std::optional for BaseRecordComponentData::m_dataset

franzpoeschel committed:
Fix the bugs found by this check

Fix the examples, the unfinished_iteratoin_test and the Coretests

franzpoeschel committed:
Don't flush in JSONIOHandlerImpl destructor

It's not needed, and it avoids weird situations while recovering from an
error.

franzpoeschel committed:
Accept undefined dataset as long as no chunks are written yet

Just ignore and skip the component in that case

franzpoeschel commented:
> This looks good! Anything on the open checks in the PR description we shall discuss tomorrow? :)

I've pushed a little fix that makes this whole PR a lot less invasive by just skipping any datasets, that have neither a dataset definition nor written chunks. That makes a flush directly after first access possible, in my opinion eliminating all remaining questions.

ax3l reviewed (approved):
:+1:
